apis/python/node/dora/__init__.pyi

```diff
@@ -0,0 +1,280 @@
+import pyarrow
+import typing
+
+@typing.final
+class Enum:
+    """Generic enumeration.
+
+Derive from this class to define new enumerations."""
+    __members__: mappingproxy = ...
+
+@typing.final
+class Node:
+    """The custom node API lets you integrate `dora` into your application.
+It allows you to retrieve input and send output in any fashion you want.
+
+Use with:
+
+```python
+from dora import Node
+
+node = Node()
+```"""
+
+    def __init__(self, /) -> None:
+        """The custom node API lets you integrate `dora` into your application.
+It allows you to retrieve input and send output in any fashion you want.
+
+Use with:
+
+```python
+from dora import Node
+
+node = Node()
+```"""
+
+    def dataflow_descriptor(self, /) -> dict:
+        """Returns the full dataflow descriptor that this node is part of.
+
+This method returns the parsed dataflow YAML file."""
+
+    def merge_external_events(self, /, subscription: Ros2Subscription) -> None:
+        """Merge an external event stream with dora main loop.
+This currently only work with ROS2."""
+
+    def next(self, /, timeout: float=None) -> PyEvent:
+        """`.next()` gives you the next input that the node has received.
+It blocks until the next event becomes available.
+You can use timeout in seconds to return if no input is available.
+It will return `None` when all senders has been dropped.
+
+```python
+event = node.next()
+```
+
+You can also iterate over the event stream with a loop
+
+```python
+for event in node:
+match event["type"]:
+case "INPUT":
+match event["id"]:
+case "image":
+```"""
+
+    def send_output(self, /, output_id: str, data: pyarrow.Array, metadata: dict=None) -> None:
+        """`send_output` send data from the node.
+
+```python
+Args:
+output_id: str,
+data: Bytes|Arrow,
+metadata: Option[Dict],
+```
+
+```python
+node.send_output("string", b"string", {"open_telemetry_context": "7632e76"})
+```"""
+
+    def __iter__(self, /) -> typing.Any:
+        """Implement iter(self)."""
+
+    def __next__(self, /) -> typing.Any:
+        """Implement next(self)."""
+
+@typing.final
+class PyEvent:
+    """Dora Event"""
+
+    def inner(self, /):...
+
+    def __getitem__(self, key: typing.Any, /) -> typing.Any:
+        """Return self[key]."""
+
+@typing.final
+class Ros2Context:
+    """ROS2 Context holding all messages definition for receiving and sending messages to ROS2.
```

Comment:
So I was thinking about putting the warnings in the Docstring. 

I understand why you would want to make things explicit and have more freedom on breaking changes. 

But IMO, I would prefer it to be a documentation statement rather than a code statement even though the 2 is obv linked. 

We also might want people to use it so that we can flag bugs and issues.

Reply:
It's just that I expect more breaking changes when we implement services and actions in the Python API, so it feels wrong to promote these types to the main module even though we will break them again soon.

Reply:
I think that the industry standard would probably to add:
- a docstring
- a warning message

Ex:
- https://github.com/pola-rs/polars/issues/12432 with implementation: https://github.com/pola-rs/polars/pull/13948
- https://github.com/dagster-io/dagster/commit/a1900c2313042f8f0a717b811ab5cade381a3789

To be honest, i have never really seen a feature marked as experimental within the function name or in the path name. And, I'm not sure if it is the most effective way to push those features.

Reply:
I guess Python is more liberal when it comes to breaking changes. I'm fine with adding a warning in the docstring for now, provided that this doesn't force us to bump the minor version on each change.

I still think that it would be a good idea to put it into a submodule to separate it from the main API of dora. It's an optional feature that we put into `libraries/extensions` in the repo for a reason. For both Rust and C++ it is disabled by default and requires enabling a cargo feature to opt-in.

Reply:
If it was a simple option, I would love to create a submodule. But the thing is that pyo3/submodule is a mess to start with. See: https://github.com/PyO3/pyo3/discussions/3591 . And even if we make it work, chances are that it is going to be quite a mental load to maintain. 


And, this does not even include auto-generated typing, which we currently do using a pulled script from https://github.com/PyO3/pyo3/issues/2454 that works but does not cover all cases and does not currently include submodules and so we will have to implement ourselves submodules typing that writes types on the submodules __init__.pyi.

And pyo3 auto-generated typing is not really making any real progress, which mean that we will have to maintain typing at the moment.


I think that we should avoid submodules and use breaking changes and warnings that are not tracked by package managers.

And revisit this in the future if we want to. 

I understand where this is coming from, and I appreciate you raising this issue. And I would have agreed if it was purely dependent on me.

Reply:
Ah well, that's unfortunate! I agree that the submodule code looks quite complex and is probably not worth the pain... Then let's merge this as-is and just add some prominent warnings about the unstability to the docs.

Reply:
So, I've just added a `warnings.warn` within `ROS2Context`. This is what `pola.rs`.

We can add more, but as most variables comes from the context i think one warning should be enough.

It will look like this:
```bash
/home/peter/Documents/work/dora/examples/python-ros2-dataflow/random_turtle.py:9: UserWarning: dora-rs Ros2 Bridge is unstable and may change at any point without it being considered a breaking change
```

Reply:
Looks good!

Maybe we should add the warning to the docs of the other Ros2 types as well? E.g. the node, publisher, and subscriber types?

Reply:
Ok sure!